packages/react-combobox/src/components/Option/useOption.tsx

```diff
@@ -62,28 +69,30 @@ export const useOption_unstable = (props: OptionProps, ref: React.Ref<HTMLElemen
       return;
     }
 
-    key && onOptionClick(event, { key, id: optionId, value: optionValue });
+    onOptionClick(event, { id, value: optionValue });
     props.onClick?.(event);
   };
 
   // register option data with context
   React.useEffect(() => {
-    if (key && optionId) {
-      return registerOption({ key, id: optionId, value: optionValue });
+    if (id && optionRef.current) {
+      const unregister = registerOption({ id, value: optionValue }, optionRef.current);
+
+      return () => unregister(id);
     }
-  }, [registerOption, optionId, key, optionValue]);
+  }, [registerOption, id, optionValue]);
```

Comment:
If we will not use `id` in `unregister` (see another comment), we can simplify it:

```ts
  React.useEffect(() => {
    if (id && optionRef.current) {
      return registerOption({ id, value: optionValue }, optionRef.current);
    }
  }, [registerOption, id, optionValue]);
```